Jetpack CP: After Jetpack plugin is installed and activated, make sure Jetpack sync is finished before considering that installation is completed.

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

This is the final part for the "Jetpack CP: Install Jetpack from app" feature: #4865.
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
After the app is able to install and activate Jetpack plugin, Jetpack will automatically do a sync to update the site's various options. We're especially interested in the update of the site's `is_woocommerce_active` option value, as that's what the app checks to determine whether the Woo site is a fully connected Jetpack site, or a Jetpack CP site (for example in Site Switch area—more info on this on p91TBi-6EK-p2)

**This PR adds the sync waiting/checking functionality.**

After that, the PR also adds various usages of the functionality, in three separate places:
1. During installation process, after installing and activating the plugin, **now the app checks for sync status.** (it will be under the "Connecting" steps in the UI).
2. If installation or activation process failed, merchants are given a button to open wp-admin in a browser and install Jetpack manually. Once merchants return from the browser back to the app, we assume that they're done with the installation, and so this PR also **adds sync checking after they return to the app.** 
3. During installation process, **if sync check fails, merchants are given a button to manually sync check again.**

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Create a new .org site with WooCommerce, but no Jetpack (Jurassic Ninja works for this),
2. Install WooComerce Payments plugin, activate it and go through steps to connect to your WordPress.com account. Skip the Stripe account afterward and just return to wp-admin,
3. Install Jetpack plugin to the site, but **don't activate it yet**
4. Start Woo Android app and select the site created in step 1,
5. Once the site is loaded, tap the Jetpack installation banner at the bottom,
6. Go through the installation steps, until it's finished ("Done" button is shown in the installation dialog),
7. Check back in wp-admin and confirm that Jetpack is now installed and activated.

### Known issue
After tapping "Done" button on step 7, the screen returns to the stats page but the Jetpack banner is still shown (even though the site now already fully connected using Jetpack-the-plugin). I'll make a new PR shortly to tackle that.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Add: variable to check whether merchant is returning from wp-admin

hafizrahman committed:
Merge branch 'hafiz/jetpack-install-success-or-fail' into hafiz/jetpack-install-handle-return-from-wpadmin

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallProgressDialog.kt

hafizrahman committed:
Add: check Jetpack sync status after installation and activation.

hafizrahman committed:
Fix: spacing.

hafizrahman committed:
Add: Use the connection checking function in the dialog.

This adds manual checking to the "Try again" button, as well as do a connection check after merchant returns from installing in wp-admin.

hafizrahman committed:
Fix: Hide groups correctly, and delete unneeded XML.

hafizrahman committed:
Add: Proper UI handling during connection retry and failure.

hafizrahman committed:
Refactor: use fetchWooCommerceSites()

Previously the code uses fetchWooCommerceSite(), but this function does not freshly fetch information for Jetpack CP site, so it can't be used for this purpose. fetchWooCommerceSites() on the other hand can fetch.

hafizrahman committed:
Add: Unit test related to sync functionality.

Also simplifies a comment in isJetpackConnectedAfterInstallation()

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/PluginRepository.kt:
```diff
@@ -161,6 +165,26 @@ class PluginRepository @Inject constructor(
+                    return true
+                } else {
+                    attempt++
+                    delay(SYNC_CHECK_DELAY)
```
During testing, usually the Jetpack sync happens quite fast, but here we're adding a bit of wait time before checking again just in case.

hafizrahman committed:
Fix: Set sync retry wait time from 1 to 3 seconds as initially planned.

hafizrahman committed:
Refactor: Move Jetpack sync checking out of repository and into vm

The thinking is that checking whether Jetpack is synced or not should not be part of the more general, plugin-related repository.

anitaa1990 commented:
@JorgeMucientes , let me know if you are taking a look at this PR today 😄 Otherwise, I can take a look at it in my evening 😊

JorgeMucientes commented:
@anitaa1990 Sure! I'll take a look at this in a bit! 🙂

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModel.kt:
```diff
@@ -80,12 +86,37 @@ class JetpackInstallViewModel @Inject constructor(
+            val result = wooCommerceStore.fetchWooCommerceSites()
+            val sites = result.model
+            if (sites != null) {
+                val syncedSite = sites.first { it.siteId == selectedSite.get().siteId }
```
I was able to make the app crash twice during the jetpack activation flow. I believe there is a race condition somewhere. When `wooCommerceStore.fetchWooCommerceSites()` returns, the resulting `sites` do not contain any matching `siteId` thus the method `sites.first { it.siteId == selectedSite.get().siteId }` throws the following exception:

```
2022-01-18 15:19:01.609 2899-2899/com.woocommerce.android.prealpha E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android.prealpha, PID: 2899
    java.util.NoSuchElementException: Collection contains no element matching the predicate.
        at com.woocommerce.android.ui.jetpack.JetpackInstallViewModel.isJetpackConnectedAfterInstallation(JetpackInstallViewModel.kt:152)
        at com.woocommerce.android.ui.jetpack.JetpackInstallViewModel.access$isJetpackConnectedAfterInstallation(JetpackInstallViewModel.kt:21)
        at com.woocommerce.android.ui.jetpack.JetpackInstallViewModel$isJetpackConnectedAfterInstallation$1.invokeSuspend(Unknown Source:14)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:234)
```

https://user-images.githubusercontent.com/2663464/149970130-621b2a3a-382c-4cfd-878f-a9762c6d1820.mp4

I was able to reproduce this crash several times, although it does not happen always. I think it's related to how fast the backend updates the site plugin status, to include the new site inside the list returned from `wooCommerceStore.fetchWooCommerceSites()` request. 

I don't know about the details on how this sync works, but some possible solution could be using this:
``` kotlin 
val syncedSite = sites.firstOrNull { it.siteId == selectedSite.get().siteId }
                if (syncedSite?.hasWooCommerce == true) {
```
And the retry if the first attempt did not work. I'm missing a lot of context so there maybe better solutions to this. 

PS: Even if the flow crashed the state of the jetpack plugin was properly transitioned to `active` in `wp-admin` and the next time you opened the app it all works as expected and jetpack is active

hichamboushaba replied:
@JorgeMucientes this is interesting, you didn't happen to keep a track of the Flipper network logs during the crash, I would love to take a look at the `/me/sites` response on this case.

But I agree with the suggested solution 👍.

hafizrahman replied:
Thanks! That's interesting and something I haven't ran into before. Do you also happen to have a lot of test sites?

In the meanwhile I've added the null check and retry.

JorgeMucientes replied:
@hichamboushaba Unfortunately I did not check the logs from Flipper yesterday. I tested it again today a couple times until I was able to reproduce it again. `me/sites` response was successful and returned a long list of my sites (101 to be exact) right before the crash. Do you want to see anything in particular? I can share the response  though Slack if you want to check it out. 

@hafizrahman No, I don't have a lot of test sites, before yesterday I had like 5, now I have 11 as I've created several to test this feature. 

With the current fix code looks good and ready to merge, although there is a small Detekt issue to be fixed @hafizrahman 
`	ImportOrdering - [] at /home/circleci/project/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModel.kt:3:1
` 

So feel free to merge it once that is fixed 🙂

hichamboushaba replied:
> Do you want to see anything in particular? I can share the response though Slack if you want to check it out.

@JorgeMucientes thanks for retesting this, yes, if you can share the response and the name of the site you were testing with, it would be great, just to confirm this is a backend glitch that happens during the sync, and that we are not missing anything.

JorgeMucientes commented on WooCommerce/src/main/res/layout/dialog_jetpack_install_progress.xml:
```diff
@@ -118,11 +118,11 @@
                 android:layout_height="@dimen/image_minor_60"
                 android:importantForAccessibility="no"
-                android:src="@drawable/ic_progress_circle_loading"
+                android:src="@drawable/ic_progress_circle_start"
```
Very minor, just in case you are not aware of there are this [naming conventions](https://github.com/woocommerce/woocommerce-android/blob/trunk/CONVENTION.md#view-ids) for view ids. Apparently we should stop using snake_case and use cameCase with the view type as suffix.

JorgeMucientes commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/jetpack/JetpackInstallViewModelTest.kt:
```diff
@@ -100,12 +119,34 @@ class JetpackInstallViewModelTest : BaseUnitTest() {
     }
+
+    @Test
+    fun `when connecting is failed, then set failed state`() = testBlocking {
```
Nitpick, I'd try to add more descriptive names here. Is not always easy though, but maybe this could be something like: 
`Given site hasWooCommerce false, when connecting jetpack, final state is Failed` 

The idea is to try to be as specific as possible on what conditions/scenario are you trying to test.

JorgeMucientes commented:
@hafizrahman Nice one! Just left a few comments of nitpicks and one blocking issue that requires some attention 🙂

hafizrahman committed:
Fix: Check for null in case site is not available yet.

There seems to be a possible race condition where the site is not yet available within the result of `fetchWooCommerceSites()`. For more details, see https://github.com/woocommerce/woocommerce-android/pull/5609#discussion_r786891870

hafizrahman committed:
Refactor: view names to match camelCase convention.

hafizrahman committed:
Merge branch 'trunk' into hafiz/jetpack-install-sync

hafizrahman committed:
Fix: Lint error.

hafizrahman committed:
Merge branch 'trunk' into hafiz/jetpack-install-sync

hafizrahman committed:
Add: update `selectedSite` after it syncs properly.

hafizrahman committed:
Add: Hide banner on My Store screen after successful installation.

JorgeMucientes commented:
Thanks for addressing all the issues. Nice job @hafizrahman! Merging
